src/main.rs

```diff
@@ -0,0 +1,192 @@
+use std::collections::HashMap;
+use std::time::Duration;
+
+use clap::Parser;
+
+use consul_external_dns::hetzner_dns;
+use tokio::time::sleep;
+
+use consul_external_dns::config::{Config, DnsProvider};
+use consul_external_dns::consul::{ConsulClient, DnsRecord};
+use consul_external_dns::dns_trait::DnsProviderTrait;
+use tokio_util::sync::CancellationToken;
+
+#[tokio::main]
+async fn main() {
+    let cancel = CancellationToken::new();
+    println!("Starting up Consul External DNS");
+
+    println!("=> Parsing configuration");
+    let config = Config::try_parse().expect("===> failed to parse configuration");
+    println!("===> parsed successfully");
+
+    let dns_provider: Box<dyn DnsProviderTrait> = match config.clone().dns_provider {
+        DnsProvider::Hetzner(config) => Box::new(hetzner_dns::HetznerDns { config }),
+    };
+
+    // Initialize Consul Client
+    println!("=> Creating Consul client");
+    let consul_client = create_consul_client(&config).await;
+    println!("===> created Consul client successfully");
+
+    // Create Consul session
+    println!("=> Creating Consul session");
+    let consul_session = consul_client
+        .create_session(cancel.clone())
+        .await
+        .expect("===> failed to create Consul session");
+    let session_id = consul_session.session_id;
+    println!("===> created Consul session successfully");
+
+    // Acquire Lock
+    println!("=> Acquiring Consul lock");
+    if let Err(e) = consul_client.acquire_lock(session_id).await {
+        eprintln!("===> failed to acquire Consul lock: {}", e);
+        return;
+    }
+    println!("===> acquired Consul lock successfully");
+
+    process_dns_records(consul_client, dns_provider, cancel).await;
+
+    consul_session
+        .join_handle
+        .await
+        .expect("Failed to join Consul session handler task");
+}
+
+async fn create_consul_client(config: &Config) -> ConsulClient {
+    loop {
+        match ConsulClient::new(
```

Comment:
`ConsulClient::new` can *only* fail as a result of the URL joining failing, and if it fails once then it will never be able to succeed again, so doing this in a loop doesn't help and so I'd suggest just calling `ConsulClient::new` in main.